search.py

```diff
@@ -572,94 +569,74 @@ def LRTA_cost(self, s, a, s1, H):
-        child1.string = parent1.string[0:split] + parent2.string[
-                                                  split:in_str_len]
-        child2.string = parent2.string[0:split] + parent1.string[
-                                                  split:in_str_len]
-        offspring.append(child1)
-        offspring.append(child2)
-
-    individuals.extend(offspring)
-    return individuals
-
-
-def mutation(individuals, in_str_len):
-    for individual in individuals:
-
-        for idx, param in enumerate(individual.string):
-            if random.uniform(0.0, 1.0) <= 0.1:
-                individual.string = individual.string[0:idx] \
-                                    + random.choice(string.ascii_letters) \
-                                    + individual.string[idx + 1:in_str_len]
-
-    return individuals
+def genetic_search(problem, fitness_fn, ngen=1000, pmut=0.1, n=20):
+    """Call genetic_algorithm on the appropriate parts of a problem.
+    This requires the problem to have states that can mate and mutate,
+    plus a value method that scores states."""
+    
+    # NOTE: This is not tested and might not work.
+    # TODO: Use this function to make Problems work with genetic_algorithm.
+    
+    s = problem.initial_state
+    states = [problem.result(s, a) for a in problem.actions(s)]
+    random.shuffle(states)
+    return genetic_algorithm(states[:n], problem.value, ngen, pmut)
+
+
+def genetic_algorithm(population, fitness_fn, gene_pool=['0', '1'], f_thres=None, ngen=1000, pmut=0.1):
+    """[Figure 4.8]"""
+    for i in range(ngen):
+        new_population = []
+        fitnesses = map(fitness_fn, population)
+        random_selection = weighted_sampler(population, fitnesses)
+        for j in range(len(population)):
+            x = random_selection()
+            y = random_selection()
+            child = reproduce(x, y)
+            if random.uniform(0, 1) < pmut:
+                child = mutate(child, gene_pool)
+            new_population.append(child)
+
+        population = new_population
+
+        if f_thres:
+            fittest_individual = argmax(population, key=fitness_fn)
+            if fitness_fn(fittest_individual) >= f_thres:
+                return fittest_individual
+
+    return argmax(population, key=fitness_fn)
+
+
+def init_population(pop_number, gene_pool, state_length):
+    """Initializes population for genetic algorithm
+    pop_number  :  Number of individuals in population
+    gene_pool   :  List of possible values for individuals
+                   (char only)
+    state_length:  The length of each individual"""
+    g = len(gene_pool)
+    population = []
+    for i in range(pop_number):
+        new_individual = ''.join([gene_pool[random.randrange(0, g)]
+                                  for j in range(state_length)])
+        population.append(new_individual)
+
+    return population
+
+
+def reproduce(x, y):
+    n = len(x)
+    c = random.randrange(0, n)
```

Comment:
Using randrange(0, n) makes the reproduction unfair since we can return unaltered `y` (when c is 0) but not `x`. The issue is very minor but makes the output biased towards `y` nonetheless.